src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1654,6 +1667,29 @@ public override BoundNode VisitCall(BoundCall node)
             }
 
             return builder == null ? default : builder.ToImmutableAndFree();
+
+            AttributeAnnotations removeMisusedNotNullWhenSense(ParameterSymbol parameter, AttributeAnnotations annotations, bool sense)
+            {
+                var whenSense = sense ? AttributeAnnotations.NotNullWhenTrue : AttributeAnnotations.NotNullWhenFalse;
+                var whenNotSense = sense ? AttributeAnnotations.NotNullWhenFalse : AttributeAnnotations.NotNullWhenTrue;
+
+                // NotNullWhenSense (without NotNullWhenNotSense) must be applied on a bool-returning member
+                if ((annotations & whenSense) != 0 &&
+                    (annotations & whenNotSense) == 0 &&
+                    parameter.ContainingSymbol.GetTypeOrReturnType().SpecialType != SpecialType.System_Boolean)
+                {
+                    annotations &= ~whenSense;
+                }
+
+                // NotNullWhenSense must be applied to a reference or unconstrained generic type
+                if ((annotations & whenSense) != 0 &&
+                    (parameter.Type?.IsValueType != false || parameter.IsParams))
```

Comment:
>parameter.IsParams [](start = 61, length = 18)

It feels like checking ```parameter.IsParams``` is not sufficient here. What are you trying to accomplish?

Reply:
Putting `[EnsuresNotNullWhenFalse]` on a `params` parameter is a misuse of the attribute, so I'm dropping the corresponding annotation.
In `bool M([EnsuresNotNullWhenFalse] params object[] x)` the attribute is ignored.

---
In reply to: [191568463](https://github.com/dotnet/roslyn/pull/26828#discussion_r191568463) [](ancestors = 191568463)